### PR TITLE
New bundle plan changes + logs

### PR DESCRIPTION
**Pull request by parishahla**
Please check out the modifications, related to :
1. Prompting bundle plan selection for app and database.
2. General error messages + bundle plan error messages.
3. Regex pattern checks for app, database and disk name. (It's the exact regex that API returned)
Definitely check the timeout error.
Please ignore the trailing commas and all prettier settings. Let me know if it's problematic in any way.

### CODE REVIEW

**Comment by parishahla** [comment]
New logging functionality
Users can now specify human-readable time formats with the since flag. It's more like Docker logs (fetching logs from the very beginning). 
Logs are only retrieved within the allowed scope of the users' bundle plan type.

**Comment by mhemrg** [comment]
By bundle plans, we actually mean: `feature bundle plans`. How can we clarify it?

File: src/commands/app/create.ts [hunk 1]
```diff
@@ -117,7 +120,51 @@ export default class AppCreate extends Command {
         }
       }
 
-      this.error(`Could not create the app. Please try again.`);
+      this.error(`Error: Unable to Create App
+        Please try the following steps:
+        1. Check your internet connection.
+        2. Ensure you have enough balance.
+        3. Try again later.
+        4. If you still have problems, please contact support by submitting a ticket at https://console.liara.ir/tickets.`);
```
Comment (mhemrg) [inline src/commands/app/create.ts:128]: Putting a dot at the end of links can be problematic:

```suggestion
        4. If you still have problems, please contact support by submitting a ticket at https://console.liara.ir/tickets`);
```

File: src/commands/app/logs.ts [hunk 2]
```diff
@@ -74,70 +76,113 @@ export default class AppLogs extends Command {
...
+
+    try {
+      const url = `v2/projects/${appName}/logs`;
+      const data = await this.got(url, {
+        searchParams: {
+          start: since,
+          direction: 'forward',
+        },
+        timeout: {
+          request: 5000,
```
Comment (eyeamnoob) [inline src/commands/app/logs.ts:151]:
```suggestion
          request: 10_000,
```

let's increase timeout

File: src/commands/app/logs.ts [hunk 3]
```diff
@@ -74,70 +76,113 @@ export default class AppLogs extends Command {
...
+        ).render(),
+      );
+      process.exit(2);
+    }
 
-      try {
-        const data = await this.got(
-          `v2/projects/${project}/logs?start=${since}`,
-        ).json<ILog>();
+    let lastLogUnix: number | null = null;
```
Comment (eyeamnoob) [inline src/commands/app/logs.ts:113]:
```suggestion
    let lastLogUnix: number = 0;
```

File: src/commands/app/logs.ts [hunk 4]
```diff
@@ -74,70 +76,113 @@ export default class AppLogs extends Command {
...
-          console.error(new Errors.CLIError(message).render());
-          process.exit(2);
-        }
-
-        this.debug(error.stack);
-      }
+    while (true) {
+      // Maybe find a better way of handling "lastLogUnix ?? 0"
+      const logs = await this.fetchLogs(
+        Math.max(start, lastLogUnix ?? 0),
```
Comment (eyeamnoob) [inline src/commands/app/logs.ts:118]:
```suggestion
        Math.max(start, lastLogUnix),
```

**Comment by eyeamnoob** [comment]
Isn't nice to show bundle plan in `liara app ls` command?

File: src/commands/app/create.ts [hunk 5]
```diff
@@ -117,7 +120,51 @@ export default class AppCreate extends Command {
...
+
+    try {
+      const { plans } = await this.got('v1/me').json<{ plans: any }>();
+
+      this.spinner.stop();
+
+      const { bundlePlan } = (await inquirer.prompt({
+        name: 'bundlePlan',
+        type: 'list',
+        message: 'Please select a plan:',
```
Comment (eyeamnoob) [inline src/commands/app/create.ts:143]:
```suggestion
        message: 'Please select a plan:',
```

The term 'plan' might refer to hardware plans. Could we come up with a more specific name for it?

**Review by eyeamnoob — COMMENTED** [review]
It seems OK.

**Comment by eyeamnoob** [comment]
I forgot to mention that, user should be able to pass bundle plan id using flag.

**Comment by parishahla** [comment]
Screenshot from 2024-06-17 15-13-21
Also keep in mind that this should be handled.

File: src/commands/db/create.ts [hunk 6]
```diff
@@ -19,11 +19,15 @@ export default class Create extends Command {
       description: 'name of your database',
     }),
     plan: Flags.string({
-      description: 'plan',
+      description:
+        'plan options are: ir-small, ir-medium, standard-base, standard-pro, pro, pro-plus, small-g2, medium-g2, standard-base-g2, standard-pro-g2, pro-g2, pro-plus-g2',
```
Comment (eyeamnoob) [inline src/commands/db/create.ts:23]: Same goes to this.

**Review by eyeamnoob — COMMENTED** [review]
Alles ist gut :)
tout est bon :)

**Comment by eyeamnoob** [comment]
> Screenshot from 2024-06-17 15-13-21 Also keep in mind that this should be handled.

It got fixed.

File: src/commands/app/create.ts [hunk 7]
```diff
@@ -20,7 +20,11 @@ export default class AppCreate extends Command {
       description: 'platform',
     }),
     plan: Flags.string({
-      description: 'plan',
+      description:
+        'plan options: free, ir-mini, ir-small, ir-medium, standard-base, standard-plus, pro, pro-plus, mini-g2, small-g2, medium-g2, standard-base-g2, standard-plus-g2, pro-g2, pro-plus-g2',
```
[the remark this answers is not in the document]
Comment (parishahla) [reply src/commands/app/create.ts:24]: This was brought back by API. Same for database.
image

**Comment by parishahla** [comment]
> It got fixed.

This piece of code could not be added. Because of the whole "error.response.body" situation.

File: src/commands/app/create.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (eyeamnoob) [reply src/commands/app/create.ts:24]: I don't think we create project or database on legacy infra, especially using CLI.
@mhemrg Any ideas?

**Comment by eyeamnoob** [comment]
> > It got fixed.
> 
> This piece of code could not be added. Because of the whole "error.response.body" situation.

What is the problem?

**Comment by parishahla** [comment]
> What is the problem?

The problem is that it's just a temporary solution. Not efficient in the long run.

**Comment by parishahla** [comment]
And please consider the fact that each time a different error response message is returned from API. The new code I added is not gonna cover all scenarios.

**Review by eyeamnoob — COMMENTED** [review]
Why don't you just simply check bundle plan and plan on front-end?
Do we ned API to tell us what is wrong, or we already know it?

**Comment by eyeamnoob** [comment]
Please mark any reviews that are resolved, as `Resolve conversation`.

**Comment by parishahla** [comment]
> Why don't you just simply check bundle plan and plan on front-end?
> Do we ned API to tell us what is wrong, or we already know it?

So this devides into "flag" and "prompt". There will be two different ways of handling it. 

It's a good idea to check it on front end when user passes the flag option.

But when it comes to prompting it, the front will become such a mess. Not efficient.
Backed should handle sending only relevant data. So that the "promptBundlePlan" function will only have one mission, destructuring and showing the options to user.

**Review by eyeamnoob — COMMENTED** [review]
Great.

File: src/commands/app/create.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (mhemrg) [reply src/commands/app/create.ts:25]: Yes. We should only keep plans ending with `-g2` + the `free` plan.

[the remark this answers is not in the document]
Comment (mhemrg) [reply src/commands/app/create.ts:25]: We already have a `plan ls` command for these cases. We don't need to mention our plans in the descriptions. Plans might change in the future.

**Comment by eyeamnoob** [comment]
One more thing, please make sure that `plan ls` command shows bundle plans as well.

**Comment by parishahla** [comment]
> It seems OK.

**Comment by parishahla** [comment]
> It seems OK.



> One more thing, please make sure that `plan ls` command shows bundle plans as well.

image
OK but we gotta fix it first.

**Review by eyeamnoob — COMMENTED** [review]
Reviewed.

**Comment by parishahla** [comment]
Please ignore the little experiment I had here.

**Comment by parishahla** [comment]
Whilst I'm on this new one, the solution to "@" at the beginning of directory name is still WIP. In a less priority though.

File: src/commands/disk/create.ts [hunk 8]
```diff
@@ -36,30 +42,47 @@ export default class DiskCreate extends Command {
...
+    } = await this.got(`v1/projects/${app}`).json<IProjectDetailsResponse>();
+
     try {
       await this.got.post(`v1/projects/${app}/disks`, { json: { name, size } });
       this.log(`Disk ${name} created.`);
     } catch (error) {
       debug(error.message);
 
+      const err = parseJSON(error.response.body);
+      console.log('err', err);
```
Comment (mhemrg) [inline src/commands/disk/create.ts:56]: Please use `this.debug` instead of `console.log`.

**Pull request merged**